Commerce docs improvements

lukeholder opened the pull request.

mattstein committed:
Spiff up Subscriptions wording.

mattstein committed:
Revise Shipping Methods and add links.

mattstein committed:
Move tip, revise heading.

Both options would require a plugin or module; the choice is shipping adjuster vs. shipping method.

mattstein committed:
Revise Extensibility.

Tighten up + link to Events.

mattstein committed:
Include event class.

mattstein committed:
Clarify that options are saved only when rule is applied.

mattstein committed:
Include ShippingMethods and Event classes in example.

mattstein committed:
Revise Events.

- Use `$event` rather than `$e` to be more explicit.
- Match language of Craft’s service docs.
- Use `// …` to consistently stand for BYO code.
- Document PdfEvent properties.

mattstein committed:
Minor edits to Commonly Used Variables.

mattstein committed:
Minor updates for Add to Cart.

mattstein committed:
Minor edits to Cart page.

mattstein committed:
Minor updates to Coupon Codes.

mattstein committed:
Minor updates for craft.commerce.carts.cart page.

mattstein committed:
Minor improvements for Custom Order Statuses.

mattstein committed:
`Control Panel` → `control panel`.

mattstein committed:
Remove line break.

mattstein committed:
Minor updates + new examples for Customer Address Management.

mattstein committed:
Attempt to clarify Customer Address Management warning.

mattstein committed:
Improvements to Customers.

mattstein committed:
Minor updates to Requirements.

mattstein committed:
Minor updates to Saving Payment Sources.

Consolidate and emphasize gateway support, clarify approaches with distinct headings. (h3 instead of h2 so sidebar nav isn’t gross.)

mattstein committed:
Trivial updates to Gateway Configuration.

mattstein committed:
Update Craft version requirement, refine wording, mention console install.

mattstein committed:
Minor edits + add copy and symlink bash examples.

mattstein committed:
Merge branch 'v3' into v3-edits

# Conflicts:
#	requirements.md

mattstein committed:
Move warning up.

mattstein committed:
Minor wording+formatting improvements for Discounts.

mattstein committed:
Fix *new* typo!

mattstein committed:
Add ”Validating addresses” section.

lukeholder commented on customers.md:
```diff
@@ -1,41 +1,41 @@
+Commerce will consolidate customers by email address. A person may checkout as a guest customer with the same email multiple times, and a new guest customer will be created each time. After an order is completed, customers are consolidated so there is only ever one customer per email address.
 
-After a user is registered and verified, whenever they log in, all guest orders that used the email address belonging to that user are transferred to the logged in user. This means, even if a registered user makes an anonymous/guest order, the next time they log in, the order will appear in their order history, as long as they used the same email address.
+Once a user is registered and verified, orders on their email address will be consolidated whenever they log in. Any guest orders with the user’s email address are transferred to the user. As a result, even future anonymous/guest orders with that email address will appear in the user’s order history.
```
This has changed in Commerce 3 and we haven‘t updated this.

How it works is the following:

All registered users in the system have an associated customer record. Then when a logged in user is building a cart on the front end, the customerId of the order/cart is always the customerId belonging to that user.

If a guest customer creates an order, a new guest customer is created for that order. When they complete that order, if there was a registered user with the same email address as the order, the order is moved to belong to the customerId of the registered user.
If there was not a registered user with the same email, but their was a previous guest customer with the same email address, the order is moved to belong to that single guest customer.

Everything now happens on order completion.

The above system means that guest customers that had their orders move to another customer, are now orphaned, and we garbage collect them (delete them) often.

let me know if you need more clarification @mattstein

mattstein replied:
Thanks Luke! This makes sense to me except for one part:

> All registered users in the system have an associated customer record.

Are you saying that a) every User now inherently has an associated Customer by default, or that b) every visitor that’s created a store account has a Customer record with an associated Craft user (as it was before)?

It seems like you’ve mostly described a change to the Customer consolidation strategy that’s now orphan-and-garbage-collect, and that the moment of tidying up is now on order completion and not login. Is that accurate?

lukeholder replied:
In Commerce 3, anytime a Craft user element is saved, we make sure a related customer record exists for that user, and create one for that user if there wasn't one.

In the Commerce 3 migration (developers moving from commerce 2) we generated a customer record for every Craft user element that didnt have one.


> It seems like you’ve mostly described a change to the Customer consolidation strategy that’s now orphan-and-garbage-collect, and that the moment of tidying up is now on order completion and not login. Is that accurate?

Exactly.

mattstein replied:
Cool, then it sounds like it’s _a)_, where a tight User<->Customer relationship is a careful design choice in v3. And the Craft User has a Customer record even if they never created an order.

This is all making sense, thanks for clarifying!

mattstein committed:
Clarify front end validation vs. server side, link to Craft docs.

mattstein committed:
Explain how to get+use validation errrors.

mattstein committed:
Minor updates for Donations page.

mattstein committed:
Update customer-user relationship description for v3.

mattstein committed:
Modest updates to Estimate Cart Addresses page.

Commented the Twig example and tried to fit it into view without horizontal scrolling, clarified `isEstimated` flag + purpose.

mattstein committed:
Updates to Making Payments.

Slightly revised language, attempted to simplify custom markup example by getting rid of unnecessary classes, element attributes, and translation filters.

mattstein committed:
Updates to Order Status Emails.

Also clarify `System Settings` → `General Settings` navigation path when relevant.

mattstein committed:
Various updates for Payment Currencies page.

Attempted to clarify and consolidate phrasing in several places, added to Twig examples and briefly documented the `stripZeros` parameter.

mattstein committed:
Reformatted Twig Filters page for (hopefully) improved readability.

mattstein committed:
Update formatting.

Let’s be more consistent with https://docs.craftcms.com/v3/dev/filters.html.

mattstein committed:
Minor edits + reformatting of Payment Form Models.

mattstein committed:
Minor edits + cleanup for Product Types.

mattstein committed:
Minor edits + `.all()` for Products Fields page.

mattstein committed:
Partial edits to Sales.

mattstein committed:
Minor phrasing + daring formatting updates for Payment Gateways.

mattstein committed:
Remaining updates for Sales.

- Adjusted heading levels to distinguish Price Effect section from Sale Application.
- Added examples for each Price Effect and edited for brevity.

mattstein committed:
Wording + formatting improvements for Customers.

Also attempted to gently re-use the phrase `registered user` referring to a customer with an associated user account.

mattstein committed:
Updates + reorganization for Shipping.

- Added two screenshots for shipping method rules.
- Changed instances of `zip code` → `ZIP code`.
- Restructured + merged Shipping headings to improve clarity.

mattstein committed:
Partial Tax edits, `ZIP code` → `zip code`.

mattstein committed:
Remaining edits for Tax page.

Tax Categories and Basic Examples are still pretty underwhelming.

mattstein committed:
Edits to Subscription Templates page.

Mostly for brevity and to improve formatting.

mattstein committed:
`controller action` → `form action`

mattstein committed:
More `controller action` → `form action`.

mattstein committed:
Edits to Update Cart Addresses.

Mostly for brevity and consistency of examples.

mattstein committed:
Add jQuery AJAX example.

mattstein committed:
Add adjuster-swapping example.

Also tidy up the examples with `use` lines and four-space indents.

mattstein commented:
@lukeholder I can’t designate you as a reviewer since I guess this is your PR, but I’ve finished going through all the pages and I’m ready for whatever feedback you’ve got!

mattstein committed:
`craft.commerce.countriesList` → `craft.commerce.countries.getAllCountriesAsList()`

lukeholder commented on adding-to-and-updating-the-cart.md:
````diff
@@ -118,28 +119,28 @@ You can add multiple purchasables to the cart in an update cart form. You supply
 ```
 
-While using multi-add the same rules apply for updating a quantity vs adding to cart, based on the uniquessness of the options `signature` and `purchasableId`.
+While using multi-add the same rules apply for updating a quantity vs. adding to cart, based on the uniquessness of the options `signature` and `purchasableId`.
````
The attribute on the line item is called `optionsSignature` which then hash of the `options` on the line item.

lukeholder commented on adding-to-and-updating-the-cart.md:
```diff
@@ -118,28 +119,28 @@ You can add multiple purchasables to the cart in an update cart form. You supply
+To update a line item, submit a `lineItems` form array, where each item’s array key is the line item ID.
 
-Example:
+Example using `LINE_ITEM_ID` as a placeholder for the ID of the item you’d like to edit:
```
Lets get rid of the `LINE_ITEM_ID` in all the examples, and just add a loop over light items within the example and use `{{ item.id }}`

lukeholder commented on customer-address-management.md:
````diff
@@ -4,54 +4,117 @@ When a customer checks out with a new address, the address is added to their add
+{% endif %}
+```
+
+For a complete template example that outputs individual field validation errors, see [shop/\_includes/addresses/form.html](https://github.com/craftcms/commerce/blob/develop/templates/shop/_includes/addresses/form.html) in the [example templates](example-templates.md).
````
is the `\` a typo in `shop/\_includes/addresses/form.html`?

lukeholder commented on twig-filters.md:
```diff
@@ -1,44 +1,59 @@
+| `currency`   | string |         | yes      | Valid, three-character ISO payment currency set up in Commerce → Store Settings → Payment Currencies.                                            |
+| `convert`    | bool   | `false` | no       | Should the amount passed to this filter be converted to `currency`’s exchange rate?                                                              |
+| `format`     | bool   | `true`  | no       | Should the amount passed to this filter be formatted according to `currency`? (Typically adds currency symbol and thousands+decimal separators.) |
+| `stripZeros` | bool   | `false` | no       | Should trailing zeros (`.00`) be stripped from a formatted number?                                                                               |
```
Tables are sooo much better! Good job.
